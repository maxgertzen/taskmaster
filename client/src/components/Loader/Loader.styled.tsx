import styled from '@emotion/styled';

export const StyledLoaderContainer = styled('div')<{ paddingTop: number }>(
  ({ theme, paddingTop }) => ({
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'column',
    width: '100%',
    paddingTop: theme.spacing(paddingTop),
  })
);

export const StyledLoaderLogo = styled('img')(({ theme }) => ({
  display: 'block',
  margin: 'auto',
  width: theme.spacing(10),
  maxWidth: '100%',
}));
