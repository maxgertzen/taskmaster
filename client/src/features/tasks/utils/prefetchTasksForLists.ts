import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
