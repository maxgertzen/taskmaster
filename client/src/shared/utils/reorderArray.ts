export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
