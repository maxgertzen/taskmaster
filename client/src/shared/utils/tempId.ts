export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
