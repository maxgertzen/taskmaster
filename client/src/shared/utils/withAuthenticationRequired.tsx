import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
