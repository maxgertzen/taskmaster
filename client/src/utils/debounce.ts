export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
