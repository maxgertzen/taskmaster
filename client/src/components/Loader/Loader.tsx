import { FC, useEffect, useState } from 'react';

import { useThemeStore } from '../../store/themeStore';
import { Title } from '../Title/Title';

import { StyledLoaderContainer, StyledLoaderLogo } from './Loader.styled';

interface LoaderProps {
  paddingTop?: number;
}

export const Loader: FC<LoaderProps> = ({ paddingTop = 0 }) => {
  const theme = useThemeStore((state) => state.theme);

  const [src, setSrc] = useState<string | null>(null);

  useEffect(() => {
    const loadGif = async () => {
      if (theme === 'dark') {
        const darkModeLoader = await import(
          '../../assets/sprites/loader-dm.gif'
        );
        setSrc(darkModeLoader.default);
      } else {
        const lightModeLoader = await import(
          '../../assets/sprites/loader-lm.gif'
        );
        setSrc(lightModeLoader.default);
      }
    };

    loadGif();
  }, [theme]);

  if (!src) return <Title variant='h4'>Loading...</Title>;

  return (
    <StyledLoaderContainer paddingTop={paddingTop}>
      <Title variant='h6'>Calculating...</Title>
      <StyledLoaderLogo src={src} alt='Loading' />
    </StyledLoaderContainer>
  );
};
