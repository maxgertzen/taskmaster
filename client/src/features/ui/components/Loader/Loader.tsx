import { FC, useEffect, useState } from 'react';

import { useThemeStore } from '@/shared/store/themeStore';

import { Title } from '..';

import { StyledLoaderContainer, StyledLoaderLogo } from './Loader.styled';

interface LoaderProps {
  paddingTop?: number;
}

export const Loader: FC<LoaderProps> = ({ paddingTop = 0 }) => {
  const theme = useThemeStore((state) => state.theme);

  const [src, setSrc] = useState<string | null>(null);

  useEffect(() => {
    const loadGif = async () => {
      if (theme === 'dark') {
        const darkModeLoader = await import(
          '@/shared/assets/animations/loader-dm.gif'
        );
        setSrc(darkModeLoader.default);
      } else {
        const lightModeLoader = await import(
          '@/shared/assets/animations/loader-lm.gif'
        );
        setSrc(lightModeLoader.default);
      }
    };

    loadGif();
  }, [theme]);

  if (!src) return <Title variant='h4'>Loading...</Title>;

  return (
    <StyledLoaderContainer paddingTop={paddingTop}>
      <Title variant='h6'>Calculating...</Title>
      <StyledLoaderLogo src={src} alt='Loading' />
    </StyledLoaderContainer>
  );
};
