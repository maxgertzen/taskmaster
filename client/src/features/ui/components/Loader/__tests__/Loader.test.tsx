import { beforeEach, describe, expect, it, vi } from 'vitest';

import {
  mockThemeStore,
  mockThemeStoreMode,
} from '@/tests/__mocks__/storeMocks';
import { render, screen } from '@/tests/utils';

import { Loader } from '../Loader';

vi.mock('@/shared/store/themeStore', () => ({
  useThemeStore: mockThemeStore,
}));

vi.mock('@/shared/assets/animations/loader-dm.gif', () => ({
  default: 'dark-mode-loader.gif',
}));

vi.mock('@/shared/assets/animations/loader-lm.gif', () => ({
  default: 'light-mode-loader.gif',
}));

describe('Loader', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe.each([
    ['light', 'light-mode-loader.gif'],
    ['dark', 'dark-mode-loader.gif'],
  ])('in %s theme', (theme, expectedSrc) => {
    beforeEach(() => {
      mockThemeStoreMode.mockReturnValue(theme);
    });

    it('renders correctly', async () => {
      const { container } = render(<Loader />);
      expect(container).toMatchSnapshot();
    });

    it('renders with custom padding', async () => {
      const { container } = render(<Loader paddingTop={20} />);
      expect(container).toMatchSnapshot();
    });

    it('loads correct gif for theme', async () => {
      render(<Loader />);
      const image = await screen.findByAltText('Loading');
      expect(image).toHaveAttribute('src', expectedSrc);
    });
  });

  describe('Loading State', () => {
    it('shows loading text while gif is loading', () => {
      render(<Loader />);
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('shows calculating text after gif loads', async () => {
      render(<Loader />);
      expect(await screen.findByText('Calculating...')).toBeInTheDocument();
    });
  });
});
