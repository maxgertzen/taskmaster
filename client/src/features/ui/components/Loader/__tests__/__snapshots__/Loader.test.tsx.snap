// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loader > in dark theme > renders correctly 1`] = `
<div>
  <h4
    css="[object Object]"
  >
    Loading...
  </h4>
</div>
`;

exports[`Loader > in dark theme > renders with custom padding 1`] = `
<div>
  <h4
    css="[object Object]"
  >
    Loading...
  </h4>
</div>
`;

exports[`Loader > in light theme > renders correctly 1`] = `
<div>
  <h4
    css="[object Object]"
  >
    Loading...
  </h4>
</div>
`;

exports[`Loader > in light theme > renders with custom padding 1`] = `
<div>
  <h4
    css="[object Object]"
  >
    Loading...
  </h4>
</div>
`;
